src/06_ffi.md

```diff
@@ -0,0 +1,395 @@
+
+Rust approach to interfacing with other programs relies on a strong
+compatibility with C. However, this boundary is by its very nature `unsafe`.
+
+Functions that are marked `extern` are made compatible with C code at the
+compilation. They may be called from C code with any parameter.
+The exact syntax is `extern "<ABI>"` where ABI is a calling convention and
+depends on the target platform. The default one is `C` which corresponds to
+a standard C calling convention on the target platform.
+
+```rust
+// export a C-compatible function
+#[no_mangle]
+pub extern "C" fn mylib_f(param: u32) -> i32 {
+    if param == 0xCAFEBABE { 0 } else { -1 }
+}
+```
+
+For the function `mylib_f` to be accessible with the same name, the function
+must also be annotated with the `#[no_mangle]` attribute.
+
+Conversely, one can call C functions from Rust if they are declared in an
+`extern` block:
+
+```rust
+// import an external function from libc
+extern "C" {
+    fn abs(args: i32) -> i32;
+}
+
+fn main() {
+    let x = -1;
+    println!("{} {}\n", x, unsafe { abs(x) });
+}
+```
+
+> **Note**
+>
+> Any foreign function imported in Rust through an `extern` block is
+> **automatically `unsafe`**. That is why, any call to a foreign function
+> must be done from an `unsafe` context.
+
+`extern` blocks may also contain foreign global variable declarations prefixed
+with the `static` keyword:
+
+```rust
+//! A direct way to access environment variables (on Unix).
+//! Should not be used! Not thread safe, have a look at `std::env`!
+
+extern {
+    // Libc global variable
+    #[link_name = "environ"]
+    static libc_environ: *const *const std::os::raw::c_char;
+}
+
+fn main() {
+    let mut next = unsafe { libc_environ };
+    while !next.is_null() && !unsafe { *next }.is_null() {
+        let env = unsafe { std::ffi::CStr::from_ptr(*next) }
+            .to_str()
+            .unwrap_or("<invalid>");
+        println!("{}", env);
+        next = unsafe { next.offset(1) };
+    }
+}
+```
+
+## Data layout
+
+Rust provides no short or long term guarantees with respect to how the data is
+laid out in the memory. The only way to make data compatible with a foreign
+language is through explicit use of a C-compatible data layout with the `repr`
+attribute. For instance, the following Rust types:
+
+```rust
+#[repr(C)]
+struct Data {
+    a: u32,
+    b: u16,
+    c: u64,
+}
+#[repr(C, packed)]
+struct PackedData {
+    a: u32,
+    b: u16,
+    c: u64,
+}
+```
+
+are compatible with the following C types:
+
+```c
+struct Data {
+    uint32_t a;
+    uint16_t b;
+    uint64_t c;
+};
+__attribute__((packed))
+struct PackedData {
+    uint32_t a;
+    uint16_t b;
+    uint64_t c;
+}
+```
+
+> ### Rule {{#check FFI-CTYPE | Use only C-compatible types in FFI}}
+>
+> In a secure Rust development, only C-compatible types must be used as
+> parameter or return type of imported or exported functions and as types of
+> imported or exported global variables.
```

Comment:
Additional mention of a not required but extremely advisable requirement: no `drop` glue for the involved types, _i.e._, avoid, or at least justify, the usage of non-`Copy` types

Reply:
Very good point. TODO